### PR TITLE
Formatted text field2

### PR DESCRIPTION
**Pull request by noodlemoodle**
- model textformatter class used to format textfield input
- move formatting logic out of formattedtextfield to numericcodeformatter
- add suffixformatter
- update formattedstring

update calling code https://github.com/nanoPayinc/NANOPAY/pull/13398

### CODE REVIEW

File: src/foam/u2/FormattedTextField.js [hunk 1]
```diff
@@ -39,7 +39,8 @@ foam.CLASS({
 
   properties: [
     {
-      class: 'Array',
+      class: 'FObjectProperty',
+      of: 'foam.u2.TextFormatter',
       name: 'formatter',
       documentation: `
         Array of integers and strings of delimiters used to format the input
```
Comment (sarthak-marwaha) [inline src/foam/u2/FormattedTextField.js]: Update documentation

**Review by sarthak-marwaha — COMMENTED** [review]
Everything else LGTM

File: src/foam/u2/TextFormatter.js [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        { name: 'end', type: 'Int' }
+      ],
+      documentation: `
+        Configure properties for delete, computes and returns a new selection range to
+        delete from
+      `
+    },
+    {
+      name: 'formatData',
+      args: [
```
Comment (kgrgreer) [inline src/foam/u2/TextFormatter.js]: Okay to keep it this way, but you could just do:
args: 'String data, String old, Int startingPos, Int endPos, Boolean isDelete'

File: src/foam/u2/TextFormatter.js [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    {
+      name: 'reset',
+      documentation: `
+        Optional method to Reset state variables on initialization, data change, and formatData calls
+      `
+    },
+    {
+      name: 'buildJavaGetFormatted',
+      args: [
+        { name: 'cls', type: 'String' },
```
Comment (kgrgreer) [inline src/foam/u2/TextFormatter.js]: same

File: src/foam/u2/TextFormatter.js [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+    function getMaxLength() {
+      return this.maxLength + this.formatter.length;
+    },
+
+    function formatData(data, old, startingPos, endPos, isDelete) {
+      // handle inserting data over the formatter portion
+      if ( ! isDelete && old && ! data.endsWith(this.formatter) ) {
+        var oldVal = old;
+        while (  old && data && data.length > startingPos && data.slice(-1) === old.slice(-1) ) {
```
Comment (kgrgreer) [inline src/foam/u2/TextFormatter.js]: remove extra space after (

**Comment by noodlemoodle** [comment]
addressed, please re-review, thanks @kgrgreer

**Pull request merged**